gwcs/wcs.py

```diff
@@ -737,80 +748,53 @@ def to_fits_sip(self, bounding_box, max_error=None, max_rms=None, degree=None,
-        #             Pix2Sky_TAN() | RotateNative2Celestial(crval1, crval2, 180))
-        # transform2 = ((Shift(-crpix1) & Shift(-crpix2)) | Mapping([0,1,0,1]) |
-        #               (fit_poly_x & fit_poly_y) | Pix2Sky_TAN() |
-        #               RotateNative2Celestial(crval1, crval2, 180))
         return hdr
 
 
-def fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
+def _fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
     """
     Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
     outputs (x, y)
     """
     llsqfitter = LinearLSQFitter()
     # The case of one pass with the specified polynomial degree
-    if degree is not None:
-        poly_x = Polynomial2D(degree=degree)
-        poly_y = Polynomial2D(degree=degree)
+    if degree is None:
+        deglist = range(2,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
         fit_poly_x = llsqfitter(poly_x, u, v, x)
         fit_poly_y = llsqfitter(poly_y, u, v, y)
-        max_resid, rms = compute_distance_residual(x, y,
-                                                   fit_poly_x(u, v), fit_poly_y(u, v))
-    else:
-        cond = True
-        degree = 2
-        while cond:
-            poly_x = Polynomial2D(degree=degree)
-            poly_y = Polynomial2D(degree=degree)
-            fit_poly_x = llsqfitter(poly_x, u, v, x)
-            fit_poly_y = llsqfitter(poly_y, u, v, y)
-            max_resid, rms = compute_distance_residual(x, y,
-                                                       fit_poly_x(u, v), fit_poly_y(u, v))
+        max_resid, rms = _compute_distance_residual(x, y,
+                                                    fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
+        if ((max_error is not None and max_resid < max_error and max_rms is None)
+             or (max_error is not None and max_resid < max_error and max_rms is not None
+                 and rms < max_rms)
+             or (max_error is None and max_rms is not None and rms > max_rms)):
             if verbose:
-                print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')
-            if ((max_error is not None and max_resid < max_error and max_rms is None)
-                 or (max_error is not None and max_resid < max_error and max_rms is not None
-                     and rms < max_rms)
-                 or (max_error is None and max_rms is not None and rms > max_rms)):
-                cond = False
-                if verbose:
-                    print('terminating condition met')
-                break
-            else:
-                degree += 1
-                if degree == 10:
-                    raise RuntimeError("Fit conditions not met")
+                print('terminating condition met')
```

Comment:
The reason I think `verbose` is not necessary is that if `'terminating condition IS NOT met'` then users will get a beautiful `RuntimeError` notification. Not getting it should be a good enough indication that things worked out. Same goes for `print(f'Degree = {degree}, max_resid = {max_resid}, rms = {rms}')`.

With regard to this `verbose` switch... Should you `print` the message or use `logging`?

In addition, I think a more "interesting" approach would be to record the error in the header, similarly to how `'CPERR'` and `'CQERR'` are currently recorded. For example, you could use `AB_ERR` and `ABP_ERR` to store the fit error in a file header. (I wish `SIP` paper would specify something like this).

Reply:
`SIPFERR` and `SIPIERR` for forward/inverse SIP errors??? :)

Reply:
verbose is useful for seeing the progress of the errors with degree.

And we could include rms errors as well. I'm happy to load it with a whole mess of them. But what is in the SIP example is the size of the correction so this is different than that. In the context of SIP it isn't obvious to those only familiar with SIP what this error means (it is the error relative to GWCS, so it is only useful to those that understand that it came from GWCS. How do we make that clear?

Reply:
Sounds like there's value in having the `verbose` flag. It's optional, so I don't see harm in having it.

Reply:
OK then. But should it use `print()` or `logging`?

Reply:
logging is not currently enabled in gwcs and is out of scope for this PR. I'd say use print for now.